remove dynamic initialization warning (#13913)

jjsjann123 opened the pull request:
removed assignment in default constructor.
removed static shared memory and used dynamic shared memory.

jjsjann123 committed:
remove dynamic initialization warning (#13913)

  removed assignment in default constructor.
  removed static shared memory and used dynamic shared memory.

soumith commented on aten/src/THC/THCReduce.cuh:
```diff
@@ -158,8 +158,8 @@ __global__ void kernelReduceNoncontigDim_shared
-  __shared__ AccT local_reduce[THC_NONCONTIG_REDUCE_BLOCK_SIZE];
-  AccT* shmem = &local_reduce[threadIdx.x + threadIdx.y*blockDim.x];
+  extern __shared__ char local_reduce[];
+  AccT* shmem = &(reinterpret_cast<AccT*>(local_reduce))[threadIdx.x + threadIdx.y*blockDim.x];
```
are there any issues if we dont check at compile-time the size of the shared memory statically? if older archs dont have `THC_NONCONTIG_REDUCE_BLOCK_SIZE` sized shared memory per block, compilation will fail right?

jjsjann123 replied:
I switched to dynamic shared memory earlier to avoid some other warning about initialization with static shared memory.
Now changing the variable to `char` instead of `AccT`, I don't really need dynamic shared memory.

For some strange reason, I'm having issue with misaligned address after switching back to static shared memory. Let me dig it a little bit more here.

ngimel replied:
Both for static and dynamic shared memory you should be taking alignment into account with `__align__` https://stackoverflow.com/questions/27570552/templated-cuda-kernel-with-dynamic-shared-memory

ngimel replied:
@soumith no, unfortunately it will happily compile, and it will be a runtime error.

jjsjann123 replied:
Cleaning the constructor and switching back to static shared memory does the trick. I'll switch back to static so we can have a compile time other than a runtime error.
Will push once the build/test finishes.

jjsjann123 committed:
switching back to static shared memory to catch compile time error of insufficient shared memory size

ezyang commented:
gcc7 failure is fake, don't worry about it

jjsjann123 commented on aten/src/THC/THCReduce.cuh:
```diff
@@ -159,7 +159,7 @@ __global__ void kernelReduceNoncontigDim_shared
   __shared__ int isLastBlockDone;
   __shared__ AccT local_reduce[THC_NONCONTIG_REDUCE_BLOCK_SIZE];
-  AccT* shmem = &local_reduce[threadIdx.x + threadIdx.y*blockDim.x];
+  AccT* shmem = &(local_reduce)[threadIdx.x + threadIdx.y*blockDim.x];
```
Sorry I was adding then removing the static_cast and somehow left the parenthesis there

jjsjann123 commented on aten/src/THC/THCTensorMathReduce.cuh:
```diff
@@ -27,6 +27,9 @@ struct WelfordData {
   __host__ __device__ WelfordData() {
+  }
+
+  __host__ __device__ void reset() {
```
Will do

jjsjann123 committed:
addressing review comments:
1. removing extra parenthesis
2. adding comment for splitting initialization from default constructor
